docs/permission/permission-framework-for-plugin-authors.md

```diff
@@ -0,0 +1,508 @@
++   Todo,
++   undefined
++ >();
+
++ export const isOwner = createTodoListPermissionRule({
++   name: 'IS_OWNER',
++   description: 'Should allow only if the todo belongs to the user',
++   apply: (resource, userId) => {
++     return resource.author === userId;
++   },
++   toQuery: userId => {
++     throw new Error('toQuery not implemented');
++   },
++ });
+
++ export const rules = [isOwner];
+```
+
+`makeCreatePermissionRule` is a helper used to ensure that rules created for this plugin use consistent types for the resource and query.
+
+We have created a new `isOwner` rule, which is going to be automatically used by the permission framework whenever a conditional response is returned in response to an authorized request with an attached `resourceRef`.
+Specifically, the `apply` function is used to understand whether the passed resource should be authorized or not.
+
+Let's skip the `toQuery` function for now.
+
+Now, let's create the new endpoint by editing `plugins/todo-list-backend/src/service/router.ts`:
+
+```diff
+- import { todosListCreate, todosListUpdate } from './permissions';
++ import { todosListCreate, todosListUpdate, TODOS_LIST_RESOURCE_TYPE } from './permissions';
++ import { rules } from './rules;
+
+  export async function createRouter(
+    options: RouterOptions,
+  ): Promise<express.Router> {
+    const { logger, identity, permissions } = options;
+
++   const permissionIntegrationRouter = createPermissionIntegrationRouter({
++     getResources: async resourceRefs => {
++       return resourceRefs.map(getTodo);
++     },
++     resourceType: TODOS_LIST_RESOURCE_TYPE,
++     rules,
++   });
+
+    const router = Router();
+    router.use(express.json());
+
++   router.use(permissionIntegrationRouter);
+
+    router.post('/todos', async (req, res) => {
+```
+
+## Authorize the `GET /todos` endpoint
+
+Authorizing the `GET /todos` is similar to the update endpoint: whenever a `GET /todos` request is received, only the items that the user has created, should be returned.
+
+As in the previous case, the permission policy can't take the decision itself, meaning that a conditional decision should be returned.
+Also, here we don't have a `resourceRef` but a list of resources.
+
+Potentially, something like this could be done:
+
+```diff
+    router.get('/todos', async (req, res) => {
+      const token = IdentityClient.getBearerToken(req.header('authorization'));
+
+-     res.json(getAll())
++     const items = getAll();
++     const decisions = await permissions.authorize(
++       items.map(({ id }) => ({ permission: todosListRead, resourceRef: id })),
++     );
+
++     const filteredItems = decisions.filter(
++       decision => decision.result === AuthorizeResult.ALLOW,
++     );
++     res.json(filteredItems);
+    });
+```
+
+This should do the trick. However, this approach has a downside. It requires to retrieve all the elements upfront and authorize them one by one.
```

Comment:
```suggestion
This should do the trick. However, this approach has a downside. It would force us to retrieve all the elements upfront and authorize them one by one, and mean that concerns like pagination that are handled by the data source today also need to move into the plugin.
```